Crashes with Android 14 and 15 devices using BioMini Slim 3

Trying to use the 0.4.1 release on Android 14 or Android 15 devices leads to a form crash and return to ODK Collect main page. 

Using Android 13 seems to work fine. 

Tested on the following devices with two different generic USB to USB-C adapters 

Pixel 7a Android 15 = Crashes to main page of ODK Collect
Galaxy Tab S8 Android 14 = Crashes to main page of ODK Collect
Redmi Note 11  Android 13 = Working

It looks like this is a problem in the BioMini scanner code:

```
java.lang.IllegalArgumentException: uk.ac.lshtm.keppel.android: Targeting U+ (version 34 and above) disallows creating or retrieving a PendingIntent with FLAG_MUTABLE, an implicit Intent within and without FLAG_NO_CREATE and FLAG_ALLOW_UNSAFE_IMPLICIT_INTENT for security reasons. To retrieve an already existing PendingIntent, use FLAG_NO_CREATE, however, to create a new PendingIntent with an implicit Intent use FLAG_IMMUTABLE.
                                                                                                    	at android.os.Parcel.createExceptionOrNull(Parcel.java:3246)
                                                                                                    	at android.os.Parcel.createException(Parcel.java:3226)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:3209)
                                                                                                    	at android.os.Parcel.readException(Parcel.java:3151)
                                                                                                    	at android.app.IActivityManager$Stub$Proxy.getIntentSenderWithFeature(IActivityManager.java:7000)
                                                                                                    	at android.app.PendingIntent.getBroadcastAsUser(PendingIntent.java:765)
                                                                                                    	at android.app.PendingIntent.getBroadcast(PendingIntent.java:748)
                                                                                                    	at uk.ac.lshtm.keppel.biomini.BioMiniScanner$mHandler$1.handleMessage(BioMiniScanner.kt:101)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:107)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:232)
                                                                                                    	at android.os.Looper.loop(Looper.java:317)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8705)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
                                                                                                    Caused by: android.os.RemoteException: Remote stack trace:
                                                                                                    	at com.android.server.am.ActivityManagerService.getIntentSenderWithFeatureAsApp(ActivityManagerService.java:5584)
                                                                                                    	at com.android.server.am.ActivityManagerService.getIntentSenderWithFeature(ActivityManagerService.java:5526)
                                                                                                    	at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:3460)
                                                                                                    	at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2812)
                                                                                                    	at android.os.Binder.execTransactInternal(Binder.java:1500)
```

--- a/Android/biomini/src/main/java/uk/ac/lshtm/keppel/biomini/BioMiniScanner.kt
+++ b/Android/biomini/src/main/java/uk/ac/lshtm/keppel/biomini/BioMiniScanner.kt
@@ -24,7 +24,7 @@ import java.util.concurrent.CountDownLatch
 private const val TAG = "KeppelBioMiniScanner"
 
 @Suppress("unused")
-class BioMiniScanner(private val context: Context) : Scanner {
+class BioMiniScanner(private val context: Context) : Scanner, BroadcastReceiver() {
 
     private val BASE_EVENT = 3000
     private val REMOVE_USB_DEVICE = BASE_EVENT + 2
@@ -48,36 +48,6 @@ class BioMiniScanner(private val context: Context) : Scanner {
     private var mFpQuality: Int? = null
 
     private var onDisconnected: (() -> Unit)? = null
-
-    private val mUsbReceiver: BroadcastReceiver = object : BroadcastReceiver() {
-        override fun onReceive(context: Context, intent: Intent) {
-            val action: String? = intent.getAction()
-            when (action) {
-                ACTION_USB_PERMISSION -> {
-                    Log.d(TAG, "ACTION_USB_PERMISSION")
-                    val hasUsbPermission: Boolean =
-                        intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)
-                    val usbDevice = mUsbDevice
-                    if (hasUsbPermission && usbDevice != null) {
-                        Log.d(TAG, usbDevice.getDeviceName() + " is acquire the usb permission. activate this device.")
-                        createBioMiniDevice()
-                    } else {
-                        Log.d(TAG, "USB permission is not granted!")
-                    }
-                }
-                UsbManager.ACTION_USB_DEVICE_ATTACHED -> {
-                    Log.d(TAG, "ACTION_USB_DEVICE_ATTACHED")
-                    findAndRequestPermission()
-                }
-                UsbManager.ACTION_USB_DEVICE_DETACHED -> {
-                    Log.d(TAG, "ACTION_USB_DEVICE_DETACHED")
-                    removeDevice()
-                    onDisconnected?.invoke()
-                }
-                else -> {}
-            }
-        }
-    }
 
     private fun removeDevice() {
         Log.d(TAG, "ACTION_USB_DEVICE_DETACHED")
@@ -135,24 +105,109 @@ class BioMiniScanner(private val context: Context) : Scanner {
         return this
     }
 
+    override fun capture(): CaptureResult? {
+        return mCurrentDevice?.let {
+            val latch = CountDownLatch(1)
+            setParameters(it)
+            doSingleCapture(latch)
+            latch.await()
+
+            val tmp = mTemplateData
+            return if (tmp?.data != null) {
+                CaptureResult(tmp.data.toHexString(), mFpQuality ?: 0)
+            } else {
+                null
+            }
+        }
+    }
+
+    override fun stopCapture() {
+        val device = mCurrentDevice
+        if (device != null) {
+            if (!device.isCapturing()) {
+                mCaptureOption.captureFuntion = IBioMiniDevice.CaptureFuntion.NONE
+                return
+            }
+            val result: Int = device.abortCapturing()
+            Log.d(TAG, "run: abortCapturing : $result")
+            if (result == 0) {
+                mCaptureOption.captureFuntion = IBioMiniDevice.CaptureFuntion.NONE
+            }
+        }
+    }
+
+    override fun disconnect() {
+        var result = 0
+        val factory = mBioMiniFactory // avoid null errors
+        if (factory != null) {
+            if (mUsbDevice != null) result = factory.removeDevice(mUsbDevice)
+            if (result == IBioMiniDevice.ErrorCode.OK.value() || result == IBioMiniDevice.ErrorCode.ERR_NO_DEVICE.value()) {
+                factory.close()
+                context.unregisterReceiver(this)
+                mUsbDevice = null
+                mCurrentDevice = null
+            }
+        }
+
+        removeDevice()
+    }
+
+    override fun onDisconnect(onDisconnected: () -> Unit) {
+        this.onDisconnected = onDisconnected
+    }
+
+    override fun onReceive(context: Context, intent: Intent) {
+        val action: String? = intent.getAction()
+        when (action) {
+            ACTION_USB_PERMISSION -> {
+                Log.d(TAG, "ACTION_USB_PERMISSION")
+                val hasUsbPermission: Boolean =
+                    intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)
+                val usbDevice = mUsbDevice
+                if (hasUsbPermission && usbDevice != null) {
+                    Log.d(
+                        TAG,
+                        usbDevice.getDeviceName() + " is acquire the usb permission. activate this device."
+                    )
+                    createBioMiniDevice()
+                } else {
+                    Log.d(TAG, "USB permission is not granted!")
+                }
+            }
+
+            UsbManager.ACTION_USB_DEVICE_ATTACHED -> {
+                Log.d(TAG, "ACTION_USB_DEVICE_ATTACHED")
+                findAndRequestPermission()
+            }
+
+            UsbManager.ACTION_USB_DEVICE_DETACHED -> {
+                Log.d(TAG, "ACTION_USB_DEVICE_DETACHED")
+                removeDevice()
+                onDisconnected?.invoke()
+            }
+
+            else -> {}
+        }
+    }
+
     private fun registerBroadcastReceiver() {
         Log.d(TAG, "start initUsbListener!")
 
         ContextCompat.registerReceiver(
             context,
-            mUsbReceiver,
+            this,
             IntentFilter(ACTION_USB_PERMISSION),
             ContextCompat.RECEIVER_NOT_EXPORTED
         )
         ContextCompat.registerReceiver(
             context,
-            mUsbReceiver,
+            this,
             IntentFilter(UsbManager.ACTION_USB_DEVICE_ATTACHED),
             ContextCompat.RECEIVER_NOT_EXPORTED
         )
         ContextCompat.registerReceiver(
             context,
-            mUsbReceiver,
+            this,
             IntentFilter(UsbManager.ACTION_USB_DEVICE_DETACHED),
             ContextCompat.RECEIVER_NOT_EXPORTED
         )
@@ -191,10 +246,6 @@ class BioMiniScanner(private val context: Context) : Scanner {
                 Log.d(TAG, "This device is not suprema device!  : " + _device.getVendorId())
             }
         }
-    }
-
-    override fun onDisconnect(onDisconnected: () -> Unit) {
-        this.onDisconnected = onDisconnected
     }
 
     private fun getCaptureCallback(latch: CountDownLatch): CaptureResponder {
@@ -272,53 +323,6 @@ class BioMiniScanner(private val context: Context) : Scanner {
                 30000 // 30 seconds
             )
         )
-    }
-
-    override fun capture(): CaptureResult? {
-        return mCurrentDevice?.let {
-            val latch = CountDownLatch(1)
-            setParameters(it)
-            doSingleCapture(latch)
-            latch.await()
-
-            val tmp = mTemplateData
-            return if (tmp?.data != null) {
-                CaptureResult(tmp.data.toHexString(), mFpQuality ?: 0)
-            } else {
-                null
-            }
-        }
-    }
-
-    override fun stopCapture() {
-        val device = mCurrentDevice
-        if (device != null) {
-            if (!device.isCapturing()) {
-                mCaptureOption.captureFuntion = IBioMiniDevice.CaptureFuntion.NONE
-                return
-            }
-            val result: Int = device.abortCapturing()
-            Log.d(TAG, "run: abortCapturing : $result")
-            if (result == 0) {
-                mCaptureOption.captureFuntion = IBioMiniDevice.CaptureFuntion.NONE
-            }
-        }
-    }
-
-    override fun disconnect() {
-        var result = 0
-        val factory = mBioMiniFactory // avoid null errors
-        if (factory != null) {
-            if (mUsbDevice != null) result = factory.removeDevice(mUsbDevice)
-            if (result == IBioMiniDevice.ErrorCode.OK.value() || result == IBioMiniDevice.ErrorCode.ERR_NO_DEVICE.value()) {
-                factory.close()
-                context.unregisterReceiver(mUsbReceiver)
-                mUsbDevice = null
-                mCurrentDevice = null
-            }
-        }
-
-        removeDevice()
     }
 }
 

